packages/core/src/routes/experience/classes/interaction-session.ts

```diff
@@ -0,0 +1,170 @@
+import { InteractionEvent, type VerificationType } from '@logto/schemas';
+import { z } from 'zod';
+
+import RequestError from '#src/errors/RequestError/index.js';
+import { type WithLogContext } from '#src/middleware/koa-audit-log.js';
+import type TenantContext from '#src/tenants/TenantContext.js';
+import assertThat from '#src/utils/assert-that.js';
+
+import type { Interaction } from '../type.js';
+
+import {
+  buildVerificationRecord,
+  verificationRecordDataGuard,
+  type VerificationRecord,
+} from './verifications/index.js';
+
+const interactionSessionResultGuard = z.object({
+  event: z.nativeEnum(InteractionEvent).optional(),
+  accountId: z.string().optional(),
+  profile: z.object({}).optional(),
+  verificationRecords: verificationRecordDataGuard.array().optional(),
+});
+
+/**
+ * InteractionSession
+ *
+ * @overview
+ * Interaction is a session that is initiated when a user starts an interaction flow with the Logto platform.
+ * This class is used to manage all the interaction data and status.
+ * @see {@link https://github.com/logto-io/rfcs | Logto RFCs} for more information about RFC 0004.
+ *
+ */
+export default class InteractionSession {
+  /**
+   * Factory method to create a new InteractionSession using context
+   */
+  static async create(ctx: WithLogContext, tenant: TenantContext) {
+    const { provider } = tenant;
+    const interactionDetails = await provider.interactionDetails(ctx.req, ctx.res);
+    return new InteractionSession(ctx, tenant, interactionDetails);
+  }
+
+  /** The interaction event for the current interaction session */
+  private interactionEvent?: InteractionEvent;
+  /** The user verification record list for the current interaction session */
+  private readonly verificationRecords: Set<VerificationRecord>;
+  /** The accountId of the user for the current interaction session. Only available once the user is identified */
+  private accountId?: string;
+  /** The user provided profile data in the current interaction session that needs to be stored to user DB */
+  private readonly profile?: Record<string, unknown>; // TODO: Fix the type
+
+  constructor(
+    private readonly ctx: WithLogContext,
+    private readonly tenant: TenantContext,
+    interactionDetails: Interaction
+  ) {
+    const { libraries, queries } = tenant;
+
+    const result = interactionSessionResultGuard.safeParse(interactionDetails.result ?? {});
+
+    assertThat(
+      result.success,
+      new RequestError({ code: 'session.interaction_not_found', status: 404 })
+    );
+
+    const { verificationRecords = [], profile, accountId, event } = result.data;
+
+    this.interactionEvent = event;
+    this.accountId = accountId; // TODO: @simeng-li replace with userId
+    this.profile = profile;
+
+    this.verificationRecords = new Set(
+      verificationRecords.map((record) => buildVerificationRecord(libraries, queries, record))
+    );
+  }
```

Comment:
It seems infeasible to use a Set for object deduplication. Should we perform deduplication based on the `type` within the objects instead?

Reply:
synced offline